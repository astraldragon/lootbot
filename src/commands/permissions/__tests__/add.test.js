jest.mock("discord-akairo");

const { Command } = require("discord-akairo");
const AddCommand = require("../add");

describe("commands/permissions/add", () => {
  beforeEach(() => {
    this.command = new AddCommand();

    this.args = {
      permission: "manage",
      role: { id: "2345" }
    };

    this.msg = {
      guild: { id: "1234" },
      client: {
        settings: {
          get: jest.fn("get").mockReturnValue({}),
          set: jest.fn("set").mockReturnValue({})
        }
      },
      channel: {
        send: jest.fn("send").mockReturnValue(Promise.resolve())
      }
    };
  });

  it("should configure properly", () => {
    expect(Command).toMatchSnapshot();
  });

  it("should update permissions if there is no matching role found", async () => {
    await this.command.exec(this.msg, this.args);

    expect(this.msg.client.settings.set).toHaveBeenCalledWith(
      this.msg.guild.id,
      "permissions",
      { manage: [this.args.role.id] }
    );

    expect(this.msg.channel.send).toHaveBeenCalledWith("Permissions updated");
  });

  it("should not update permissions if the role is already added", async () => {
    this.msg.client.settings.get.mockReturnValue({
      manage: [this.args.role.id]
    });

    await this.command.exec(this.msg, this.args);
    expect(this.msg.client.settings.set).not.toHaveBeenCalled();
  });

  it("should notify if any errors occur", async () => {
    this.msg.client.settings.get.mockImplementation(() => {
      throw new Error("test");
    });

    await this.command.exec(this.msg, this.args);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      "An error occurred updating permissions"
    );
  });
});
