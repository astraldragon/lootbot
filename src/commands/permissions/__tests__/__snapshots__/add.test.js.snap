// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands/permissions/add should configure properly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "permissions-add",
      Object {
        "aliases": Array [
          "permissions-add",
          "pa",
        ],
        "args": Array [
          Object {
            "id": "permission",
            "prompt": Object {
              "start": "What is the name of the permission? (manage)",
            },
            "type": "string",
          },
          Object {
            "id": "role",
            "prompt": Object {
              "start": "What is the role to assign the permission?",
            },
            "type": "role",
          },
        ],
        "category": "Permissions",
        "channelRestriction": "guild",
        "description": Object {
          "content": "Add role to lootbot",
          "examples": Array [
            "permissions-add manage \\"Loot Manager\\"",
          ],
          "usage": "<permission> <role>",
        },
        "options": Object {
          "permissions": Array [
            "ADMINISTRATOR",
          ],
        },
        "split": "quoted",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
